File: 328.odd-even-linked-list.cs
/*
 * @lc app=leetcode id=328 lang=csharp
 *
 * [328] Odd Even Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    public ListNode OddEvenList(ListNode head) 
    {
        if (head == null || head.next == null)
        {
            return head;
        }
        else
        {
            ListNode odd = null;
            ListNode even = null;
            ListNode tempOdd = null;
            ListNode tempEven = null;
            int count = 0;
            ListNode temp = head;
            while (temp != null)
            {
                if (count % 2 == 0)
                {
                    if (count == 0)
                    {
                        odd = temp;
                        tempOdd = temp;
                    }
                    else
                    {
                        tempOdd.next = temp;
                        tempOdd = tempOdd.next;
                    }
                }
                else
                {
                    if (count == 1)
                    {
                        even = temp;
                        tempEven = temp;
                    }
                    else
                    {
                        tempEven.next = temp;
                        tempEven = tempEven.next;
                    }
                }
                temp = temp.next;
                count++;
            }
            tempEven.next = null;
            tempOdd.next = even;
            return odd;
        }
    }
}
// @lc code=end

File: 328.odd-even-linked-list.py
#
# @lc app=leetcode id=328 lang=python3
#
# [328] Odd Even Linked List
#

# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional

class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        else:
            odd = None
            even = None
            count = 0
            temp = head
            while temp:
                if not odd:
                    if count == 0:
                        odd = temp
                        temp_odd = temp
                else:
                    if count % 2 == 0:
                        temp_odd.next = temp
                        temp_odd = temp_odd.next
                if not even:
                    if count == 1:
                        even = temp
                        temp_even = temp
                else:
                    if count % 2 == 1:
                        temp_even.next = temp
                        temp_even = temp_even.next
                temp = temp.next
                count += 1
            temp_even.next = None
            temp_odd.next = even
            return odd
# @lc code=end

